Guides/BSON.md

```diff
@@ -0,0 +1,429 @@
+```swift
+var doc: Document = [
+    "a": 5
+    "b": ObjectId()
+]
+let value: BSONValue? = doc["a"] // 5
+let intValue = (value as? BSONNumber)?.int32Value // Int32(5)
+doc["c"] = "i am a string"
+```
+This API provided a number of benefits, the principal one being the seamless integration of standard Swift types (e.g. `Int`) and driver custom ones (e.g `ObjectId`) into `Document`'s methods. It also had a few drawbacks, however. In order for `BSONValue` to be used as an existential type, it could not have `Self` or associated type requirements. This ended being a big restriction as it meant `BSONValue` could not be `Equatable`, `Hashable`, or `Codable`. Instead, all of this functionaltiy was put onto the separate wrapper type `AnyBSONValue`, which was used instead of an existential `BSONValue` in meany places in order to leverage these common protocol conformances.
+
+Another drawback is that subdocuments lterals could not be inferred and had to be explicitly casted:
+```swift
+let x: Document = [
+    "x": [
+        "y": [
+            z: 4
+        ] as Document
+    ] as Document
+]
+```
+### Required Updates
+In BSON API v2, `BSONNumber`, `BSONValue`, and `AnyBSONValue` no longer exist. They are all entirely replaced by the `BSON` enum.
+
+#### Updating `BSONValue` references
+
+Anywhere in the driver that formerly accepted or returned a `BSONValue` will now accept or return a `BSON`. Wherever `BSONValue` is used as an existential value in your application, a `BSON` will probably work as a drop-in replacement. Any casts will need to be updated to call the appropriate helper property instead.
+```swift
+func foo(x: BSONValue) -> String? {
+    guard let stringValue = x as? String else {
+        return nil
+    }
+    return "foo" + stringValue
+}
+```
+becomes:
+```swift
+func foo(x: BSON) -> String? {
+    guard let stringValue = x.stringValue else {
+        return nil
+    }
+    // or
+    guard case let .string(stringValue) = x else {
+        return nil
+    }
+    return "foo" + stringValue
+}
+```
+Similarly, `BSON`'s `Equatable` conformance can be leveraged instead of the old `bsonEquals`.
+```swift
+func foo(x: BSONValue, y: BSONValue) { 
+    if x.bsonEquals(y) { ... }
+}
+```
+becomes simply:
+```swift
+func foo(x: BSON, y: BSON) {
+    if x == y { ... }
+}
+```
+**Generic Requirement**
+Currently, there is no equivalent protocol in BSON API v2 to `BSONValue`, so if your application was using it as a generic requirement there is no alternative in the driver. You may have to implement your own similar protocol to achieve the same effect. If such a protocol would be useful to you, please file a ticket on the driver's Jira project.
```

Comment:
add link to README section letting them know how to get to the jira page quickly

Reply:
done